query_server/query/src/extension/logical/optimizer_rule/implicit_type_conversion.rs

```diff
@@ -138,41 +136,43 @@ impl OptimizerRule for ImplicitTypeConversion {
     }
 }
 
-struct DataTypeRewriter<'a> {
+struct DataTypeRewriter {
     /// input schemas
-    schemas: Vec<&'a DFSchemaRef>,
+    schema: DFSchemaRef,
 }
 
-impl<'a> DataTypeRewriter<'a> {
+impl DataTypeRewriter {
+    pub fn new(schemas: Vec<&DFSchemaRef>) -> Self {
+        let mut schema = DFSchema::empty();
+        schemas.into_iter().for_each(|s| schema.merge(s.as_ref()));
+        Self {
+            schema: Arc::new(schema),
+        }
+    }
     /// If expr is Column, get its data type; otherwise, return None
-    fn extract_column_data_type(&self, expr: &Expr) -> Option<DataType> {
+    fn extract_column_data_type(&self, expr: &Expr) -> Result<Option<DataType>> {
         if let Expr::Column(_) = expr {
-            for schema in &self.schemas {
-                if let Ok(v) = expr.get_type(schema) {
-                    return Some(v);
-                }
-            }
+            return Ok(Some(expr.get_type(&self.schema)?));
         }
-
-        None
+        Ok(None)
     }
 
     fn convert_data_type_if_necessary(
         &self,
         mut left: Box<Expr>,
         mut right: Box<Expr>,
     ) -> Result<(Box<Expr>, Box<Expr>)> {
-        let left_type = self.extract_column_data_type(&left);
-        let right_type = self.extract_column_data_type(&right);
+        let left_type = self.extract_column_data_type(&left)?;
+        let right_type = self.extract_column_data_type(&right)?;
 
         // Ensure that the left side of the op is column
         let mut reverse = false;
         let left_type = match (&left_type, &right_type) {
-            (Some(v), None) => v,
-            (None, Some(v)) => {
+            (Some(t), None) => t.clone(),
+            (None, Some(t)) => {
                 reverse = true;
                 mem::swap(&mut left, &mut right);
-                v
+                t.clone()
```

Comment:
what is the purpose of clone

Reply:
reference here:

https://github.com/apache/arrow-datafusion/blob/4c7833ebfdb2d022830bb97862e0ce36b0b3d6b1/datafusion/optimizer/src/analyzer/type_coercion.rs#L70-L99